src/java/org/apache/cassandra/db/virtual/PrimaryIdTable.java

```diff
@@ -0,0 +1,350 @@
+    private static final String COLUMN_KEY = "key";
+    private static final String COLUMN_SIZE_ESTIMATE = "size_estimate";
+    private static final String COLUMN_SSTABLES = "sstables";
+
+    private final TableMetadata metadata;
+    private final ColumnMetadata sizeEstimateColumn;
+    private final ColumnMetadata sstablesColumn;
+
+    @VisibleForTesting
+    final CopyOnWriteArrayList<Consumer<DecoratedKey>> readListener = new CopyOnWriteArrayList<>();
+
+    public PrimaryIdTable(String keyspace)
+    {
+        this.metadata = TableMetadata.builder(keyspace, NAME)
+                                     .kind(TableMetadata.Kind.VIRTUAL)
+                                     .partitioner(new LocalPartitioner(CompositeType.getInstance(UTF8Type.instance, UTF8Type.instance)))
+                                     .addPartitionKeyColumn(COLUMN_KEYSPACE_NAME, UTF8Type.instance)
+                                     .addPartitionKeyColumn(COLUMN_TABLE_NAME, UTF8Type.instance)
+                                     .addClusteringColumn(COLUMN_TOKEN_VALUE, IntegerType.instance)
+                                     .addClusteringColumn(COLUMN_KEY, UTF8Type.instance)
+                                     .addRegularColumn(COLUMN_SIZE_ESTIMATE, CounterColumnType.instance)
+                                     .addRegularColumn(COLUMN_SSTABLES, CounterColumnType.instance)
+                                     .build();
+        sizeEstimateColumn = metadata.regularColumns().getSimple(0);
+        sstablesColumn = metadata.regularColumns().getSimple(1);
+    }
+
+    @Override
+    public UnfilteredPartitionIterator select(DecoratedKey partitionKey, ClusteringIndexFilter clusteringIndexFilter, ColumnFilter columnFilter, RowFilter rowFilter)
+    {
+        if (clusteringIndexFilter.isReversed())
+            throw new InvalidRequestException(REVERSED_QUERY_ERROR);
+
+        ByteBuffer[] key = ((CompositeType) this.metadata.partitionKeyType).split(partitionKey.getKey());
+        String keyspace = UTF8Type.instance.getString(key[0]);
+        String table = UTF8Type.instance.getString(key[1]);
+
+        KeyspaceMetadata ksm = Schema.instance.getKeyspaceMetadata(keyspace);
+        if (ksm == null)
+            throw invalidRequest(KEYSPACE_NOT_EXIST_ERROR, keyspace);
+
+        TableMetadata metadata = ksm.getTableOrViewNullable(table);
+        if (metadata == null)
+            throw invalidRequest(TABLE_NOT_EXIST_ERROR, table, keyspace);
+
+        AbstractBounds<PartitionPosition> range = getBounds(metadata, clusteringIndexFilter, rowFilter);
+        return new SingletonUnfilteredPartitionIterator(select(partitionKey, metadata, clusteringIndexFilter, range));
+    }
+
+    private List<SSTableReader> getSStables(TableMetadata metadata, AbstractBounds<PartitionPosition> range)
+    {
+        return Lists.newArrayList(ColumnFamilyStore.getIfExists(metadata).getTracker().getView().liveSSTablesInBounds(range.left, range.right));
```

Comment:
if you query a table that doesn't exist, or is a virtual table, won't this blow up?  should we validate ahead of time?

Reply:
it does about 10 lines up when it checks and throws KEYSPACE_NOT_EXIST_ERROR or TABLE_NOT_EXIST_ERROR